### PR TITLE
Add clearer error message using `&str + &str`

### PR DESCRIPTION
**Pull request by mgattozzi**
This is the first part of #39018. One of the common things for new users
coming from more dynamic languages like JavaScript, Python or Ruby is to
use `+` to concatenate strings. However, this doesn't work that way in
Rust unless the first type is a `String`. This commit adds a check for
this use case and outputs a new error as well as a suggestion to guide
the user towards the desired behavior. It also adds a new test case to
test the output of the error.

### CODE REVIEW

File: src/librustc_typeck/check/op.rs [hunk 1]
```diff
@@ -238,6 +244,36 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
...
+                    let suggestion =
+                        match (codemap.span_to_snippet(lhs_expr.span),
+                                codemap.span_to_snippet(rhs_expr.span)) {
+                            (Ok(lstring), Ok(rstring)) =>
+                                format!("{}.to_string() + {}", lstring, rstring),
+                            _ => format!("<expression>")
+                        };
+                    err.span_suggestion(expr.span,
+                        &format!("Your first argument should be a `String`, not `&str`. \
+                                    Try using `to_string()`"),
```
Comment (Manishearth) [inline src/librustc_typeck/check/op.rs]: `.to_string()`

File: src/test/compile-fail/E0369.rs [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+fn main() {
+    let string = "Hello " + "World" + "!";
+    //~^ ERROR 12:18: 12:26: binary operation `+` cannot be applied to type `&'static str` [E0369]
+    //~| NOTE You can't use `+` between a `&str` and `&str`
+    //~| HELP Your first argument should be a `String`, not `&str`. Try using `to_string()`
+    //~| NOTE an implementation of `std::ops::Add` might be missing for `&'static str`
```
Comment (Manishearth) [inline src/test/compile-fail/E0369.rs]: Have a thingy for the suggestion too

**Comment by Manishearth** [comment]
Mostly looks okay? Would like to see if @jonathandturner has thoughts on this.

Feel free to try and add handling for the other forms. `String` is a struct so you can search for it.

I've been considering having lang-item-like annotations for types in the stdlib, which let diagnostics check for their existence. We do path matching extensively in clippy, and it would be useful to do something less hacky in rustc.

**Review by Manishearth — APPROVED** [review]

File: src/librustc_typeck/check/op.rs [hunk 3]
```diff
@@ -221,8 +223,12 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
                         };
 
                         if let Some(missing_trait) = missing_trait {
+                            if missing_trait == "std::ops::Add" {
+                                self.check_str_addition(expr, lhs_expr, lhs_ty,
+                                                        rhs_expr, rhs_ty_var, &mut err);
+                            }
                             span_note!(&mut err, lhs_expr.span,
-                                       "an implementation of `{}` might be missing for `{}`",
+                                        "an implementation of `{}` might be missing for `{}`",
```
Comment (bjorn3) [inline src/librustc_typeck/check/op.rs]: Please remove this extra space.

**Review by bjorn3 — CHANGES_REQUESTED** [review]

**Comment by mgattozzi** [comment]
I've updated the code with the above suggestions. @Manishearth how would I go about searching for a `String` so that I could match against it here.

**Comment by Manishearth** [comment]
This is what we do in clippy to match structs.

https://github.com/Manishearth/rust-clippy/blob/630e0ef96875e0958277674f81921bfaba30bc5f/clippy_lints/src/utils/mod.rs#L140-L179

I think we should just land as-is since that's a bit hacky. I'm filing an issue for the diagnostic-lang-item stuff.

r? @jonathandturner

**Comment by Manishearth** [comment]
Filed https://github.com/rust-lang/rust/issues/39131 about the diagnostic item

**Comment by mgattozzi** [comment]
In the meantime I think it might be good to get the diagnostic of things like `String + String` if possible but that could be a separate PR if that's better.

**Comment by Manishearth** [comment]
Actually, you can make this work without the full framework of diagnostics items; just add a `#[diagnostic_item = "String"]` annotation and then in your code check that `cx.tcx.get_attrs(adt_def.def_id)` has that attribute.

**Comment by mgattozzi** [comment]
@Manishearth using the code from clippy or just putting it in the code now as is?

**Comment by Manishearth** [comment]
No, we don't want to use the clippy code, it's hacky. Use an attribute as shown above.

**Comment by Manishearth** [comment]
Maybe `rustc_diagnostic_item`

**Comment by mgattozzi** [comment]
I used ripgrep and git grep to search the code and neither of those attributes exist in the code base so I have no basis of comparison to do it the way you've mentioned.

**Comment by Manishearth** [comment]
I'm saying you invent a new attribute :)

Annotate String with `#[rustc_diagnostic_item = "String"]`. Then, using `tcx.get_attrs(adt_def.def_id)`, you can check if your `TyAdt` has that attribute.

You will probably have to add it to libsyntax/feature_gate.rs as well, as a Gated attribute.

**Comment by Manishearth** [comment]
Anyway we can land as-is if you'd prefer too.

**Comment by jonathandturner** [comment]
I like the idea.

Can you use a src/test/ui test instead of a compile_fail test?  Just so that we can see what the final output looks like a bit better in the PR.

**Comment by mgattozzi** [comment]
@Manishearth ohhhhhh okay I think I see what you're talking about now. I can try when I'm off work.

@jonathandturner I'll switch it over to ui test when I'm off work!

File: src/librustc_typeck/check/op.rs [hunk 4]
```diff
@@ -238,6 +244,36 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
...
+                          lhs_ty: Ty<'tcx>,
+                          rhs_expr: &'gcx hir::Expr,
+                          rhs_ty_var: Ty<'tcx>,
+                          mut err: &mut errors::DiagnosticBuilder) {
+        let rhs_ty = self.check_expr_coercable_to_type(rhs_expr, rhs_ty_var);
+        if let TyRef(_, l_ty) = lhs_ty.sty {
+            if let TyRef(_, r_ty) = rhs_ty.sty {
+                if l_ty.ty.sty == TyStr && r_ty.ty.sty == TyStr {
+                    span_note!(&mut err, lhs_expr.span,
+                            "You can't use `+` between a `&str` and `&str`");
```
Comment (nrc) [inline src/librustc_typeck/check/op.rs]: We usually use the passive voice for error messages, i.e., something like "`+` can't be used to concatenate two `&str` strings" and similarly later on.

File: src/librustc_typeck/check/op.rs [hunk 5]
```diff
@@ -238,6 +244,36 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
...
+                    let codemap = self.tcx.sess.codemap();
+                    let suggestion =
+                        match (codemap.span_to_snippet(lhs_expr.span),
+                                codemap.span_to_snippet(rhs_expr.span)) {
+                            (Ok(lstring), Ok(rstring)) =>
+                                format!("{}.to_string() + {}", lstring, rstring),
+                            _ => format!("<expression>")
+                        };
+                    err.span_suggestion(expr.span,
+                        &format!("Your first argument should be a `String`, not `&str`. \
```
Comment (nrc) [inline src/librustc_typeck/check/op.rs]: It would be nice to explain *why* a String is needed, otherwise this seems a little like boilerplate.

File: src/librustc_typeck/check/op.rs [hunk 6]
```diff
@@ -238,6 +244,36 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
...
+                                    Try using to_string()"),
```
Comment (nrc) [inline src/librustc_typeck/check/op.rs]: I would prefer phrasing like "`to_string` can be used to create an owned string from a string reference", I actually prefer using `to_owned` rather than `to_string` because it better describes what is going on

File: src/test/compile-fail/E0369.rs [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+fn main() {
+    let string = "Hello " + "World" + "!";
+    //~^ ERROR 12:18: 12:26: binary operation `+` cannot be applied to type `&'static str` [E0369]
+    //~| NOTE You can't use `+` between a `&str` and `&str`
+    //~| HELP Your first argument should be a `String`, not `&str`. Try using to_string()
+    //~| SUGGESTION let string = "Hello ".to_string() + "World" + "!"
+    //~| NOTE an implementation of `std::ops::Add` might be missing for `&'static str`
```
Comment (nrc) [inline src/test/compile-fail/E0369.rs]: Would be nice if we could skip the NOTE in this case.

**Review by nrc — CHANGES_REQUESTED** [review]

File: src/test/compile-fail/E0369.rs (hunk 7 above)
Comment (jonathandturner) [inline src/test/compile-fail/E0369.rs]: @nrc - I think he's going to move to a UI test, so we won't need them (and we'll be able to see what the actual error will look like)

Comment (nrc) [inline src/test/compile-fail/E0369.rs]: I meant remove it from the error message, not from the test

Comment (mgattozzi) [inline src/test/compile-fail/E0369.rs]: Shouldn't be too hard. I have an idea of how to bypass this output.

File: src/librustc_typeck/check/op.rs (hunk 6 above)
Comment (mgattozzi) [inline src/librustc_typeck/check/op.rs]: `to_owned` or `to_string` is always a hot button topic. Now they're semantically equivalent at least. I can put `to_owned`. It's all the same to me and with the rest of the sentence you put it communicates the intent of the function call.

File: src/librustc_typeck/check/op.rs (hunk 4 above)
Comment (mgattozzi) [inline src/librustc_typeck/check/op.rs]: Consistency is good. This one is a quick fix.

File: src/librustc_typeck/check/op.rs (hunk 5 above)
Comment (mgattozzi) [inline src/librustc_typeck/check/op.rs]: Yeah I can see why it's boilerplate like. I'll put a more descriptive message.

**Comment by mgattozzi** [comment]
@nrc I've updated the wording, though I'm wondering if it could use some improvement. I didn't want to make it too verbose, but I wanted to get the point across as well. I've also added a way to bypass the NOTE you didn't want output here through some conditional logic short circuiting. It feels a little hackish and I was wondering if you had a better idea here as to what could be improved.

Also not since the test hasn't been updated yet this should fail testing.

File: src/librustc_typeck/check/op.rs [hunk 8]
```diff
@@ -223,13 +223,18 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
...
 
                         if let Some(missing_trait) = missing_trait {
-                            if missing_trait == "std::ops::Add" {
-                                self.check_str_addition(expr, lhs_expr, lhs_ty,
-                                                        rhs_expr, rhs_ty_var, &mut err);
+                            // The logic here is saying if the trait was an add
+                            // and it was an addition with &str then skip the
+                            // span note below it. Otherwise just emit the span
+                            // note
+                            if missing_trait == "std::ops::Add" &&
```
Comment (bjorn3) [inline src/librustc_typeck/check/op.rs]: This will also skip the note when the trait is not Add instead of only when Add + &str

**Review by bjorn3 — CHANGES_REQUESTED** [review]

File: src/librustc_typeck/check/op.rs (hunk 8 above)
Comment (mgattozzi) [inline src/librustc_typeck/check/op.rs]: You're right. I got the logic wrong here. I tested it just to make sure so this will need to get changed

File: src/librustc_typeck/check/op.rs [hunk 9]
```diff
@@ -238,6 +248,45 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
...
+                        match (codemap.span_to_snippet(lhs_expr.span),
+                                codemap.span_to_snippet(rhs_expr.span)) {
+                            (Ok(lstring), Ok(rstring)) =>
+                                format!("{}.to_owned() + {}", lstring, rstring),
+                            _ => format!("<expression>")
+                        };
+                    err.span_suggestion(expr.span,
+                        &format!("to_owned() can be used to create an owned `String` \
+                                  from a string reference. This allows concatenation \
+                                  since the `String` is owned."), suggestion);
```
Comment (nrc) [inline src/librustc_typeck/check/op.rs]: The second sentence doesn't quite explain why that is necessary. How about, "String concatenation appends the string on the right to the string on the left. That modifies the string on the left and may require reallocation, this requires ownership of the string on the left." It's a bit wordy, but tries to explain why you can't use a reference.

**Comment by mgattozzi** [comment]
@nrc and @jonathandturner these commits should cover both the ui test and message output that was brought up earlier.

**Comment by mgattozzi** [comment]
Looking at the raw log one failed on my test due to the output being different but it passed for all the other ones. I'm not sure why this would happen though

**Review by nrc — APPROVED** [review]

**Comment by nrc** [comment]
r = me with the test fixed. I'm not sure why that is failing, but it looks legit to me.

**Comment by mgattozzi** [comment]
@nrc looking at the logs further this seems to be the problem

```
expected stderr:
error[E0514]: found crate `std` compiled by an incompatible version of rustc
  |
  = help: please recompile that crate using this compiler (rustc 1.16.0-dev (b27d71cb1 2017-01-23))
  = note: crate `std` path #1: /home/michael/Code/rustc/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/libstd-86f66de14176b4c2.rlib compiled by "rustc 1.16.0-dev (5c73f42db 2017-01-23)"
  = note: crate `std` path #2: /home/michael/Code/rustc/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/libstd-86f66de14176b4c2.so compiled by "rustc 1.16.0-dev (5c73f42db 2017-01-23)"
  = note: crate `std` path #3: /home/michael/Code/rustc/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/libstd_unicode-b421c85d3facafd6.rlib compiled by "rustc 1.16.0-dev (5c73f42db 2017-01-23)"
  = note: crate `std` path #4: /home/michael/Code/rustc/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/libstd_shim-dac73ac8aa295bd7.rlib compiled by "rustc 1.16.0-dev (5c73f42db 2017-01-23)"

error: aborting due to previous error
```

I had this happen once or twice during my builds to test all this, though I'm not sure what caused it. Usually just a clean and rebuild fixed it. Possibly restarting the build might get rid of the error.

**Comment by mgattozzi** [comment]
@nrc I fixed the build!

**Comment by nrc** [comment]
@mgattozzi nice! Could you squash the commits please? Then r+

**Comment by mgattozzi** [comment]
@nrc squash and force push?

**Comment by nrc** [comment]
yes

**Comment by mgattozzi** [comment]
@nrc got it. Wasn't sure policy regarding force pushing. Alright it's all rebased and good to go!

**Comment by nrc** [comment]
@bors: r+

Thanks for sticking with it @mgattozzi , nice PR!

**Comment by mgattozzi** [comment]
@nrc I'm glad to have been able to contribute to the compiler itself finally! It was a joy and I'm hoping to tackle some more issues. Thanks for reviewing it!

**Comment by mgattozzi** [comment]
@nrc looks like curl messed up trying to download the rustc binary to do compilation.

**Comment by kevinmehall** [comment]
Looks like this unintentionally added files under `src/target/debug/` to git.

**Comment by mgattozzi** [comment]
@kevinmehall thanks I'll remove it

**Comment by mgattozzi** [comment]
@nrc it needs to get approved again so it can go through

**Comment by Manishearth** [comment]
Finally

**Comment by mgattozzi** [comment]
@Manishearth and @nrc thanks a lot! This great that it's finally merged!
